1.8.9/src/main/java/io/github/axolotlclient/modules/hypixel/bedwars/upgrades/BedwarsTeamUpgrades.java

```diff
@@ -37,49 +47,103 @@ public class BedwarsTeamUpgrades {
 	);
 
 	public final TeamUpgrade maniacMiner = new TieredUpgrade(
 		"haste", Pattern.compile("^\\b[A-Za-z0-9_§]{3,16}\\b purchased Maniac Miner .{1,3}\\s*$"),
-		new int[]{2, 4}, new int[]{4, 6}, new TextureInfo[]{
-		new TextureInfo("textures/gui/container/inventory.png", 2*18, 198, 256, 256, 18, 18, Color.DARK_GRAY),
-		new TextureInfo("textures/gui/container/inventory.png", 2*18, 198, 256, 256, 18, 18, Color.GRAY),
-		new TextureInfo("textures/gui/container/inventory.png", 2*18, 198, 256, 256, 18, 18),
+		new int[]{2, 4}, new int[]{4, 6}, (x, y, width, height, upgradeLevel) -> {
+		if (upgradeLevel == 1) {
+			Color color = Color.GRAY;
+			GlStateManager.color(color.getAlpha() / 255F, color.getRed() / 255F, color.getGreen() / 255F, color.getBlue() / 255F);
+		} else if (upgradeLevel == 0) {
+			Color color = Color.DARK_GRAY;
+			GlStateManager.color(color.getAlpha() / 255F, color.getRed() / 255F, color.getGreen() / 255F, color.getBlue() / 255F);
+		}
+		MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("textures/gui/container/inventory.png"));
+		DrawableHelper.drawTexture(x, y, 2*18, 198, 18, 18, 16, 16, 256, 256);
 	}
 	);
 
 	public final TeamUpgrade forge = new TieredUpgrade(
 		"forge", Pattern.compile("^\\b[A-Za-z0-9_§]{3,16}\\b purchased (?:Iron|Golden|Emerald|Molten) Forge\\s*$"),
-		new int[]{2, 4}, new int[]{4, 6}, new TextureInfo[]{
-		new TextureInfo("textures/blocks/furnace_front_off.png", 6*18, 198 + 18, 18, 18),
-		new TextureInfo("textures/blocks/furnace_front_on.png", 6*18, 198 + 18, 18, 18),
-		new TextureInfo("textures/blocks/furnace_front_on.png", 6*18, 198 + 18, 18, 18, Color.parse("#FFFF00")),
-		new TextureInfo("textures/blocks/furnace_front_on.png", 6*18, 198 + 18, 18, 18, Color.parse("#00FF00")),
-		new TextureInfo("textures/blocks/furnace_front_on.png", 6*18, 198 + 18, 18, 18, Color.parse("#FF0000")),
+		new int[]{2, 4}, new int[]{4, 6}, (x, y, width, height, upgradeLevel) -> {
+		if(upgradeLevel == 0){
+			MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("textures/block/furnace_front.png"));
+			DrawableHelper.drawTexture(x, y, 0, 0, width, height, width, height);
+		} else {
+			if(upgradeLevel == 2){
+				Color color = Color.parse("#FFFF00");
+				GlStateManager.color(color.getAlpha() / 255F, color.getRed() / 255F, color.getGreen() / 255F, color.getBlue() / 255F);
+			} else if (upgradeLevel == 3) {
+				Color color = Color.parse("#00FF00");
+				GlStateManager.color(color.getAlpha() / 255F, color.getRed() / 255F, color.getGreen() / 255F, color.getBlue() / 255F);
+			} else if (upgradeLevel == 4){
+				Color color = Color.parse("#FF0000");
+				GlStateManager.color(color.getAlpha() / 255F, color.getRed() / 255F, color.getGreen() / 255F, color.getBlue() / 255F);
+			}
+			MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("textures/block/furnace_front_on.png"));
+			DrawableHelper.drawTexture(x, y, 0, 0, width, height, width, height);
```

Comment:
For forge, protection, and maniac it may be nice to have an option that displays the tier of the upgrade on the bottom right of the icon (just like stack number). So protection 1 would display a 1, 2 display a 2...

Reply:
protection and haste are easy to tell apart (especially protection now, it looks exactly like sybillian's textures but uses the resource pack provided textures instead), for forge *maybe* (but also relatively trivial)

Reply:
added it for the forge upgrade, the other ones are easy enough to distinguish imo (but it could still be added later, it's only one line that'd need to be copied)